services/clusters-keeper/tests/unit/test_modules_ec2.py

```diff
@@ -0,0 +1,325 @@
+    assert running_instance["InstanceType"] == instance_type
+    assert "Tags" in running_instance
+    assert running_instance["Tags"] == [
+        {"Key": key, "Value": value} for key, value in tags.items()
+    ]
+
+
+async def test_start_aws_instance_is_limited_in_number_of_instances(
+    mocked_aws_server_envs: None,
+    aws_vpc_id: str,
+    aws_subnet_id: str,
+    aws_security_group_id: str,
+    aws_ami_id: str,
+    ec2_client: EC2Client,
+    clusters_keeper_ec2: ClustersKeeperEC2,
+    app_settings: ApplicationSettings,
+    faker: Faker,
+    mocker: MockerFixture,
+):
+    assert app_settings.CLUSTERS_KEEPER_EC2_ACCESS
+    assert app_settings.CLUSTERS_KEEPER_EC2_INSTANCES
+    # we have nothing running now in ec2
+    all_instances = await ec2_client.describe_instances()
+    assert not all_instances["Reservations"]
+
+    # create as many instances as we can
+    tags = faker.pydict(allowed_types=(str,))
+    startup_script = faker.pystr()
+    for _ in range(
+        app_settings.CLUSTERS_KEEPER_EC2_INSTANCES.EC2_INSTANCES_MAX_INSTANCES
+    ):
+        await clusters_keeper_ec2.start_aws_instance(
+            app_settings.CLUSTERS_KEEPER_EC2_INSTANCES,
+            faker.pystr(),
+            tags=tags,
+            startup_script=startup_script,
+            number_of_instances=1,
+        )
+
+    # now creating one more shall fail
+    with pytest.raises(Ec2TooManyInstancesError):
+        await clusters_keeper_ec2.start_aws_instance(
+            app_settings.CLUSTERS_KEEPER_EC2_INSTANCES,
+            faker.pystr(),
+            tags=tags,
+            startup_script=startup_script,
+            number_of_instances=1,
+        )
+
+
+async def test_get_instances(
+    mocked_aws_server_envs: None,
+    aws_vpc_id: str,
+    aws_subnet_id: str,
+    aws_security_group_id: str,
+    aws_ami_id: str,
+    ec2_client: EC2Client,
+    clusters_keeper_ec2: ClustersKeeperEC2,
+    app_settings: ApplicationSettings,
+    faker: Faker,
+    mocker: MockerFixture,
+):
+    assert app_settings.CLUSTERS_KEEPER_EC2_INSTANCES
+    # we have nothing running now in ec2
+    all_instances = await ec2_client.describe_instances()
+    assert not all_instances["Reservations"]
+    assert (
+        await clusters_keeper_ec2.get_instances(
+            app_settings.CLUSTERS_KEEPER_EC2_INSTANCES, {}
+        )
+        == []
+    )
+
+    # create some instance
+    instance_type = faker.pystr()
+    tags = faker.pydict(allowed_types=(str,))
+    startup_script = faker.pystr()
+    created_instances = await clusters_keeper_ec2.start_aws_instance(
+        app_settings.CLUSTERS_KEEPER_EC2_INSTANCES,
+        instance_type,
+        tags=tags,
+        startup_script=startup_script,
+        number_of_instances=1,
+    )
+    assert len(created_instances) == 1
+
+    instance_received = await clusters_keeper_ec2.get_instances(
+        app_settings.CLUSTERS_KEEPER_EC2_INSTANCES,
+        tags=tags,
+    )
```

Comment:
Q: if you have less tags in the dataset will it still return it and is it something you want?